bin/propolis-server/src/lib/vm/state_driver.rs

```diff
@@ -478,10 +489,10 @@ impl StateDriver {
         info!(self.log, "state driver launched");
 
         let final_state = if migrated_in {
-            if self.start_vm(VmStartReason::MigratedIn).await.is_ok() {
-                self.event_loop().await
-            } else {
-                InstanceState::Failed
+            match self.start_vm(VmStartReason::MigratedIn).await {
+                VmStartOutcome::Succeeded => self.event_loop().await,
+                VmStartOutcome::Failed => InstanceState::Failed,
+                VmStartOutcome::Aborted => InstanceState::Destroyed,
```

Comment:
it occurs to me that we _could_ just return `Result<(), InstanceState>` in `start_vm`...but I think `VmStartOutcome` is maybe a bit better at expressing the semantic meaning to the reader. Returning a `Result` would let `start_vm` use `?` if it `map_err`ed results to `InstanceState`s, but...i don't actually see a ton of places where we might want to do that...

Reply:
I do think we can cut back on the stuttering a little by putting a method on `VmStartOutcome` that yields the expected final state for that outcome (if there is one). That at least removes the repetition between the explicit-request and migrated-in startup paths.